README.md
# Clash Speedtest

![Golang](https://img.shields.io/github/actions/workflow/status/starudream/clash-speedtest/golang.yml?label=golang&style=for-the-badge)
![Docker](https://img.shields.io/github/actions/workflow/status/starudream/clash-speedtest/docker.yml?label=docker&style=for-the-badge)
![Release](https://img.shields.io/github/v/release/starudream/clash-speedtest?include_prereleases&sort=semver&style=for-the-badge)
![License](https://img.shields.io/github/license/starudream/clash-speedtest?style=for-the-badge)

`Clash` 节点测速

## Configure

| Variable | Type         | Default               | Description                      |
|----------|--------------|-----------------------|----------------------------------|
| DEBUG    | BOOL         | FALSE                 | show debug information           |
| URL      | STRING       | http://127.0.0.1:9090 | clash external controller url    |
| SECRET   | STRING       | -                     | clash external controller secret |
| PROXY    | STRING       | http://127.0.0.1:7890 | configuration file path          |
| TIMEOUT  | STRING       | 5s                    | timeout for http request         |
| INCLUDE  | STRING ARRAY | -                     | filter nodes by include          |
| EXCLUDE  | STRING ARRAY | -                     | filter nodes by exclude          |
| RETRY    | INT          | 3                     | retry times when failed          |
| OUTPUT   | STRING       | -                     | output directory                 |

### Docker

![Version](https://img.shields.io/docker/v/starudream/clash-speedtest?sort=semver&style=for-the-badge)
![Size](https://img.shields.io/docker/image-size/starudream/clash-speedtest?sort=semver&style=for-the-badge)
![Pull](https://img.shields.io/docker/pulls/starudream/clash-speedtest?style=for-the-badge)

```bash
docker pull starudream/clash-speedtest
```

```bash
docker run --rm \
    --name clash-speedtest \
    -e DEBUG=true \
    -e URL=http://host.docker.internal:9090 \
    -e PROXY=http://host.docker.internal:7890 \
    starudream/clash-speedtest
```

## Screenshot

![screenshot](./docs/screenshot.png)

## License

[Apache License 2.0](./LICENSE)


CHANGELOG.md
# Changelog

## [2.2.2](https://github.com/starudream/clash-speedtest/compare/v2.2.1...v2.2.2) (2023-02-03)


### Performance Improvements

* perfect ([6a76a59](https://github.com/starudream/clash-speedtest/commit/6a76a5981758f67c78cc2a1b2533c1eec847652d))

## [2.2.1](https://github.com/starudream/clash-speedtest/compare/v2.2.0...v2.2.1) (2022-12-13)


### Features

* **deps:** bump dependencies ([46e94e1](https://github.com/starudream/clash-speedtest/commit/46e94e1b8dba25ed6ecdbe72ca243b1b54802d91))

## [2.2.0](https://github.com/starudream/clash-speedtest/compare/v2.1.2...v2.2.0) (2022-12-02)


### Features

* **deps:** bump dependencies ([4f06240](https://github.com/starudream/clash-speedtest/commit/4f06240db7ba8e5b1e0948aac32cb65918bdf596))

## [2.1.2](https://github.com/starudream/clash-speedtest/compare/v2.1.1...v2.1.2) (2022-10-09)


### Performance Improvements

* perfect use go 1.19 ([87090e0](https://github.com/starudream/clash-speedtest/commit/87090e0cabefaa179ce83c7f3c36a6a4c2906fdb))

## [2.1.1](https://github.com/starudream/clash-speedtest/compare/v2.1.0...v2.1.1) (2022-09-29)


### Bug Fixes

* fix proxy type filter ([315572e](https://github.com/starudream/clash-speedtest/commit/315572ecadef4df46ccc32f62344ce471447caec))

## [2.1.0](https://github.com/starudream/clash-speedtest/compare/v2.0.0...v2.1.0) (2022-09-09)


### release

* 2.1.0 ([931d353](https://github.com/starudream/clash-speedtest/commit/931d353770db3c881c5e69b06779bde61827a2dc))

## [2.0.0](https://github.com/starudream/clash-speedtest/compare/v1.0.2...v2.0.0) (2022-08-30)


### release

* 2.0.0 ([6e612cd](https://github.com/starudream/clash-speedtest/commit/6e612cd1a5c01dd4d282f6687e12fc63dfa2660f))

## [1.0.2](https://github.com/starudream/clash-speedtest/compare/v1.0.1...v1.0.2) (2022-06-16)


### Performance Improvements

* perfect app ([47cdd18](https://github.com/starudream/clash-speedtest/commit/47cdd182ceccd406c9154cdca79606b19857ce20))

## [1.0.1](https://github.com/starudream/clash-speedtest/compare/v1.0.0...v1.0.1) (2022-06-10)


### Bug Fixes

* fix release action ([7a5213c](https://github.com/starudream/clash-speedtest/commit/7a5213c4d8c0c84269b716ab5d8410e38aac6d72))

## [1.0.0](https://github.com/starudream/clash-speedtest/compare/v0.0.2...v1.0.0) (2022-06-10)


### Features

* add release action ([d0c3dbf](https://github.com/starudream/clash-speedtest/commit/d0c3dbf47527fac9bb561208444b5898f62119f8))


### release

* 1.0.0 ([9c3868e](https://github.com/starudream/clash-speedtest/commit/9c3868ed4aae1ac9feb886247f049664b6c87f31))


cmd/main.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"net"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/olekukonko/tablewriter"
	"github.com/starudream/go-lib/app"
	"github.com/starudream/go-lib/config"
	"github.com/starudream/go-lib/log"

	"github.com/starudream/clash-speedtest/clash"
	"github.com/starudream/clash-speedtest/download"
)

func init() {
	log.Attach("app", "clash-speedtest")

	config.SetDefault("url", "http://127.0.0.1:9090")
	config.SetDefault("proxy", "http://127.0.0.1:7890")
	config.SetDefault("retry", 3)
	config.SetDefault("timeout", "5s")
}

func main() {
	app.Add(initSpeedtest)
	app.Defer(deferSpeedtest)
	err := app.OnceGo()
	if err != nil {
		log.Fatal().Msgf("app init fail: %v", err)
	}
}

var (
	smu     sync.Mutex
	mode    clash.Mode
	results []*download.Result
)

func initSpeedtest(context.Context) error {
	clashCli := clash.New(config.GetString("url"), config.GetString("secret"))
	downloadCli := download.New(config.GetString("proxy"), config.GetDuration("timeout"))

	version, err := clashCli.GetVersion()
	if err != nil {
		return err
	}

	log.Info().Msgf("clash version: %s, premium: %t", version.Version, version.Premium)

	mode, err = clashCli.GetMode()
	if err != nil {
		return err
	}

	proxies, err := clashCli.GetProxies()
	if err != nil {
		return err
	}

	global := proxies["GLOBAL"]
	if global == nil || len(global.All) == 0 {
		return fmt.Errorf("clash no global proxy")
	}

	err = clashCli.SetMode(clash.ModeGlobal)
	if err != nil {
		return err
	}

	var (
		include = config.GetStringSlice("include")
		exclude = config.GetStringSlice("exclude")
		retry   = config.GetInt("retry")
	)

	if retry <= 0 {
		retry = 1
	}

	log.Info().Msgf("include: %#v, exclude: %#v", include, exclude)

	var names []string

	for i := 0; i < len(global.All); i++ {
		proxy, exist := proxies[global.All[i]]
		if !exist {
			return fmt.Errorf("clash proxy %s not exist", global.All[i])
		}

		l := log.With().Str("name", proxy.Name).Str("type", proxy.Type).Logger()

		if len(exclude) > 0 && includeKeywords(proxy.Name, exclude) {
			l.Debug().Msgf("exclude, skip proxy")
			continue
		}

		if len(include) > 0 && !includeKeywords(proxy.Name, include) {
			l.Debug().Msgf("include, skip proxy")
			continue
		}

		switch proxy.Type {
		case "Shadowsocks", "ShadowsocksR", "Snell", "Socks5", "Http", "Vmess", "Trojan":
		case "Direct", "Reject", "Relay", "Selector", "Fallback", "URLTest", "LoadBalance":
			continue
		default:
			l.Warn().Msgf("clash type %s not support", proxy.Type)
			continue
		}

		names = append(names, proxy.Name)

		l.Debug().Msgf("add proxy")
	}

	log.Info().Msgf("speedtest count: %d", len(names))

	if len(names) == 0 {
		log.Warn().Msgf("no proxy to test")
		return nil
	}

	defer deferSpeedtest()

	for i := 0; i < len(names); i++ {
		proxy := proxies[names[i]]

		l := log.With().Int("index", i+1).Str("name", proxy.Name).Str("type", proxy.Type).Logger()

		err = clashCli.SetProxy(proxy.Name)
		if err != nil {
			return err
		}

		l.Info().Msgf("speedtest begin")

		for j := 0; j < retry; j++ {
			result, se := downloadCli.Cloudflare()
			if se != nil {
				l.Error().Msgf("speedtest error: %v", se)
				continue
			}

			result.Name = proxy.Name
			result.Type = proxy.Type

			smu.Lock()
			results = append(results, result)
			smu.Unlock()

			l.Info().
				IPAddr("ip", net.ParseIP(result.IP)).
				Str("country", result.Country).Str("city", result.City).
				Str("lat", result.Lat).Str("lng", result.Lng).
				Msgf(
					"speedtest end, took %.02fs, duration %dms, download %s, speed %s",
					result.TotalDuration.Seconds(), result.ConnectDuration.Milliseconds(), formatMeter(float64(result.Length)), formatMeter(result.BS)+"/s",
				)

			break
		}
	}

	return nil
}

func deferSpeedtest() {
	func() {
		smu.Lock()
		defer smu.Unlock()
		if len(results) > 0 {
			text := formatResult(results)
			fmt.Println(text)
			filename := filepath.Join(config.GetString("output"), fmt.Sprintf("result-%s.txt", time.Now().Format("20060102150405")))
			err := os.WriteFile(filename, []byte(text), 0644)
			if err != nil {
				log.Fatal().Msgf("write file fail: %v", err)
			}
			results = nil
		}
	}()
	func() {
		if mode != "" {
			_ = clash.New(config.GetString("url"), config.GetString("secret")).SetMode(mode)
			mode = ""
		}
	}()
}

func formatResult(results []*download.Result) string {
	bb := &bytes.Buffer{}
	tw := tablewriter.NewWriter(bb)
	tw.SetAlignment(tablewriter.ALIGN_CENTER)
	tw.SetHeader([]string{"name", "country", "type", "duration", "speed", "total"})

	total := float64(0)
	for i := 0; i < len(results); i++ {
		v := results[i]
		total += float64(v.Length)
		var color int
		if v.BS < 1024*1024 {
			color = tablewriter.FgRedColor
		} else if v.BS < 5*1024*1024 {
			color = tablewriter.FgYellowColor
		} else {
			color = tablewriter.FgGreenColor
		}
		tw.Rich(
			[]string{v.Name, v.Country, v.Type, fmt.Sprintf("%dms", v.ConnectDuration.Milliseconds()), formatMeter(v.BS) + "/s", formatMeter(float64(v.Length))},
			[]tablewriter.Colors{{tablewriter.Bold}, {}, {}, {}, {tablewriter.Bold, color}, {}},
		)
	}

	tw.SetFooter([]string{"", "", "", "", "", formatMeter(total)})
	tw.Render()
	return bb.String()
}

func formatMeter(v float64) string {
	if v < 1024 {
		return fmt.Sprintf("%.02fB", v)
	}
	v /= 1024
	if v < 1024 {
		return fmt.Sprintf("%.02fKB", v)
	}
	v /= 1024
	if v < 1024 {
		return fmt.Sprintf("%.02fMB", v)
	}
	v /= 1024
	if v < 1024 {
		return fmt.Sprintf("%.02fGB", v)
	}
	v /= 1024
	return fmt.Sprintf("%.02fTB", v)
}

func includeKeywords(v string, keywords []string) bool {
	for _, k := range keywords {
		if strings.Contains(v, k) {
			return true
		}
	}
	return false
}
